### ISSUE TITLE
Immutable properties with dynamic values

### ISSUE THREAD

**Issue by fwkoch** [3ptscience/properties#105]
This would be useful for defining one property with a value totally dependent on other properties. The behaviour would be very similar to defining and `@property`, but as a `Property` it would have additional documentation, declarative definition, possibly additional validation.

For example, instead of something like this:
```
class BuriedSource(properties.HasProperties):
    depth = properties.Float(
        doc='Depth of source'
    )
    @property
    def loc_vector(self):
        return vectormath.Vector3([0., 0., -self.depth])
```
you could do:
```
class BuriedSource(properties.HasProperties):
    depth = properties.Float(
        doc='Depth of source'
    )
    loc_vector = properties.Vector3(
        doc='Vector location of source',
        dynamic_value=lambda self: [0., 0., -self.depth]
    )
```
`loc_vector` would then be auto-documented and coerced to a Vector3 automatically.

By default, a property with `dynamic_value` defined would be otherwise immutable - the value is only calculated on `get` and never `set`. Possibly there could be another attribute like `overwritable` where the dynamic value can be permanently overwritten (at least until it is deleted).

**Comment by rowanc1** [comment]
Could we actually use these as wrappers?

```python
class BuriedSource(properties.HasProperties):
    depth = properties.Float(
        doc='Depth of source'
    )
    @properties.Vector3
    def loc_vector(self):
        """Vector location of source"""
        return vectormath.Vector3([0., 0., -self.depth])
```

Then `@loc_vector.setter` if necessary later. Perhaps this way of using the properties does not do any storage - only validation and documentation.

How is this different from the `_default` (it is dynamic, but only on the first time?) or using an `observer` on depth?

**Comment by fwkoch** [comment]
`default` is only called on construction and is independent of other property values.

An `observer` on depth could provide similar functionality. However, we still want `loc_vector` to be a property so it is documented. If we don't want the user to set it, we'd need to use `GettableProperty` and there we lose `Vector3`-specific validation and documentation. This could all be remedied by adding more text to `doc` and calling `Vector3` validation function in the observed function but that is all a bit complicated.

I like the idea of a wrapper that adds documentation (and maybe validation), but doesn't save anything to the backend.

**Comment by lheagy** [comment]
I 👍 the wrapper. I think it makes it quite clean, especially for properties that might be computed in a couple steps.

**Comment by fwkoch** [comment]
Implemented in #108
